keps/sig-cli/kubectl-commands-in-headers.md

```diff
@@ -0,0 +1,219 @@
+  - "@kow3ns"
+approvers:
+  - "@soltysh"
+  - "@seans3"
+editor: TBD
+creation-date: 2019-02-22
+last-updated: 2019-02-22
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# kubectl-commands-in-headers
+
+
+## Table of Contents
+
+A table of contents is helpful for quickly jumping to sections of a KEP and for highlighting any additional information provided beyond the standard KEP template.
+[Tools for generating][] a table of contents from markdown are available.
+
+- [Table of Contents](#table-of-contents)
+- [Release Signoff Checklist](#release-signoff-checklist)
+- [Summary](#summary)
+- [Motivation](#motivation)
+- [Goals](#goals)
+- [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+- [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Implementation History](#implementation-history)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Release Signoff Checklist
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [x] KEP approvers have set the KEP status to `implementable`
+- [x] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [x] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Requests sent to the apiserver from kubectl include http request headers with context about the kubectl command that
+created the request.  This information could be used by cluster administrators for debugging or
+to gather telemetry about how users are interacting with the cluster.
+
+
+## Motivation
+
+Kubectl generates requests sent to the apiserver for commands such as `apply`, `delete`, `edit`, `run`, however
+the context of the command for the requests is lost and unavailable to cluster administrators.  Context would be
+useful to cluster admins both for debugging the cause of requests as well as providing telemetry about how users
+are interacting with the cluster.
+
+### Goals
+
+- Allow cluster administrators to identify how requests in the logs were generated from
+  kubectl commands.
+
+### Non-Goals
+
+NA
+
+## Proposal
+
+Include in requests made from kubectl to the apiserver:
+
+- the kubectl subcommand
+- which flags were specified (but not the values)
```

Comment:
will there be an opt-in for certain cases?  I'd like to see the `kubectl patch --type` value to know which kinds of patches are actually being frequently used.

Reply:
Yeah, it might be useful to have a mechanism to indicate which flag values are sensitive or not. `-o` for example is complex because some values can be sensitive (`-o column-name=...`) and some aren't (`-o yaml`). We could improve the flag registration mechanisms that we use in kubectl to have this sort of semantic maybe?

Reply:
SGTM.  We could start with a whitelisted set of flags we will include an enumerated value for (never arbitrary values supplied by the user).  We can expand the list over time.  That was the intention for -f and -o values.

Reply:
> SGTM. We could start with a whitelisted set of flags we will include an enumerated value for (never arbitrary values supplied by the user). We can expand the list over time. That was the intention for -f and -o values.

Yeah, I think that would work well.

Reply:
Updated